feat(#225): Send sync duration in Firebase Analytics as a custom event and in Settings row

JosephSanjaya opened the pull request:
## 📱 Description
<!-- Provide a brief description of what this PR accomplishes -->

## Platform
- [x] Android
- [ ] iOS
- [ ] Games-Unity
- [ ] DevOps (AWS)
- [ ] Website
- [ ] C/Golang 

## 🎯 Type of Change
<!-- Mark the relevant option with an [x] -->
- [ ] 🐛 Bug fix
- [x] ✨ New feature
- [ ] 🔧 Refactoring (code change that neither fixes a bug nor adds a feature)
- [ ] 📚 Documentation update
- [ ] 🎨 UI/UX improvement
- [ ] ⚡ Performance improvement
- [ ] 🧪 Test addition or improvement

## 📋 Changes
- `TimeProvider`: A utility class for getting the current time, allowing for easier testing.
- `AnalyticsSource`: An interface for logging analytics events.
- `FirebaseAnalyticsSource`: An implementation of `AnalyticsSource` that uses Firebase Analytics.
- `SyncAnalyticsRepository`: A repository to manage and persist sync-related analytics data, and log events.
  - `startSync()`: Records the start time of a sync.
  - `stopSync()`: Records the stop time of a sync and accumulates the duration if the sync is paused/resumed.
  - `completeSync()`: Persists the total sync duration, end timestamp, and a unique ID. It then logs a "user_did_complete_sync" event with these details and the current block height.
  - `getLastSyncMetadata()`: Retrieves the metadata of the last completed sync.
- Unit tests for the new classes.
- Integration of `SyncAnalyticsRepository` into `SyncManager` to call the respective methods at the start, stop, and completion of the sync process.
- `FakeSharedPreferences`: A testing utility for mocking SharedPreferences.
- The `getFormatted()` method has been removed from `SyncAnalyticsRepository.SyncMetadata` and its logic is now encapsulated in the `SyncMetadata.Formatter.format()` method.
- This change improves the separation of concerns and makes the formatting logic more testable.
- The `HomeSettingDrawerSheet` has been updated to use the new `Formatter` class.
- Unit tests have been added to `SyncAnalyticsRepositoryTest` to verify the correctness of the `SyncMetadata.Formatter`.

### 🔗 Related Issues
<!-- Link any related issues using "Fixes #issue_number" or "Closes #issue_number" -->
- Fixes https://github.com/gruntsoftware/internal/issues/225

## 🧪 Tests Status
- [x] Tests ran successfully locally?
- [x] Added more tests? How many?
- [ ] Code coverage percentage of the codebase: __%

## 📸 Screenshots/Videos
<!-- Add screenshots or screen recordings of UI changes -->
<!-- Use the format below for before/after comparisons -->

| Before | After |
|--------|-------|
|<!-- Add screenshot/video of current behavior -->|<!-- Add screenshot/video of new behavior -->|

## 🎯 Reviewers
<!-- Tag specific reviewers or teams -->
@kcw-grunt, @josikie

JosephSanjaya committed:
feat(#225): Send sync duration in Firebase Analytics as a custom event and in Settings row

This commit introduces analytics tracking for the sync process.

It includes:
- `TimeProvider`: A utility class for getting the current time, allowing for easier testing.
- `AnalyticsSource`: An interface for logging analytics events.
- `FirebaseAnalyticsSource`: An implementation of `AnalyticsSource` that uses Firebase Analytics.
- `SyncAnalyticsRepository`: A repository to manage and persist sync-related analytics data, and log events.
  - `startSync()`: Records the start time of a sync.
  - `stopSync()`: Records the stop time of a sync and accumulates the duration if the sync is paused/resumed.
  - `completeSync()`: Persists the total sync duration, end timestamp, and a unique ID. It then logs a "user_did_complete_sync" event with these details and the current block height.
  - `getLastSyncMetadata()`: Retrieves the metadata of the last completed sync.
- Unit tests for the new classes.
- Integration of `SyncAnalyticsRepository` into `SyncManager` to call the respective methods at the start, stop, and completion of the sync process.
- `FakeSharedPreferences`: A testing utility for mocking SharedPreferences.

kcw-grunt commented:
👀  @JosephSanjaya

kcw-grunt commented on app/src/main/java/com/brainwallet/data/repository/SyncAnalyticsRepository.kt:
```diff
@@ -0,0 +1,84 @@
+        private const val KEY_LAST_END = "last_sync_end_timestamp"
+        private const val KEY_DURATION_MILLIS = "duration_millis"
+        private const val KEY_END_TIMESTAMP = "end_timestamp"
+        private const val KEY_BLOCK_HEIGHT = "block_height"
```
Assuming this is ok note there are core Interfaces that may make it even more succinct
The BRTransaction Class (the core C-class running the low level wallet functions) might be helpful to review.
Here is a link (@JosephSanjaya ): https://github.com/gruntsoftware/core/blob/10ae536cab23f197b60a6feac042a07dc8d61ae0/BRTransaction.h#L86

Did you already review?

JosephSanjaya replied:
Thanks for the review, currently im still on going understanding the process, i will update the PR again

JosephSanjaya replied:
Added formatter for the metadata, also change the source for `getLastBlockTimestamp` and `getCurrentBlockHeight` from BRPeer 

cc @kcw-grunt

josikie reviewed (commented):
👀👀👀

kcw-grunt reviewed (changes requested):
Looks very close @JosephSanjaya .

We do need to polish the layout so users can understand.

<img width="421" height="348" alt="Screenshot 2025-08-28 at 14 58 36" src="https://github.com/user-attachments/assets/2f92c31b-1970-467a-9488-7ade4aae5cbb" />

Figma link is here: 
https://www.figma.com/design/9IP5iHEq0jxxujWA9sg2n2/brainwallet-app-ui-ux?node-id=4872-2847&t=XZlj6ulgA55S8ckI-1

JosephSanjaya committed:
chore(#225): Move SyncMetadata formatting to Formatter class

This commit introduces a `Formatter` class within `SyncAnalyticsRepository.SyncMetadata` to handle the formatting of sync metadata.

The `getFormatted()` method has been removed from `SyncAnalyticsRepository.SyncMetadata` and its logic is now encapsulated in the `SyncMetadata.Formatter.format()` method.

This change improves the separation of concerns and makes the formatting logic more testable.

The `HomeSettingDrawerSheet` has been updated to use the new `Formatter` class.

Unit tests have been added to `SyncAnalyticsRepositoryTest` to verify the correctness of the `SyncMetadata.Formatter`.

kcw-grunt reviewed (approved):
LGTM
